### PR TITLE
`Exam mode`: Fix exam exercise name shown to students in breadcrumb

### PR DESCRIPTION
**Pull request by tobias-lippert**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context
Follow-up to #6873 to make the breadcrumbs consistent with the title shown for the exam exercise to students.

### Description
We now distinguish on the server side if the exercise name for the breadcrumb should be shown to a student or someone who is at least a tutor. In the former case, we show the exercise group name in the breadcrumb if it's an exam exercise, otherwise the normal exercise title. In the latter case, we always show the exercise title because for tutors the exercises should be distinguishable.
@Strohgelaender thanks for fixing the jpql query.

### Steps for Testing
- 1 Student
- 1 Tutor or user with more privileges
- 1 Exam

1. Participate in the exam and submit the exam
2. Grade the exam (also add an example solution for at least one exercise) and check that the exercise and not the exercise group name is shown in the breadcrumb.
3. Review your results as student and verify that the breadcrumb text for the exercise matches the exercise title you see
4. Review the example solution as student and verify the same
5. check the exercise names on the exam summary page at the top where the points are listed.

### Exam mode testing
Part of the normal testing steps

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [exercise.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5189/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/shared/exercise/exercise.service.ts.html) | 92.38% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |Comment|
|------------|--------------:|---------------------:|-------------------:|
| [ExerciseRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5189/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/ExerciseRepository.html) | 76% | ✅ | cannot be directly covered|
| [TitleCacheEvictionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5189/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/TitleCacheEvictionService.html) | 85% |  | |
### Screenshots

image

### CODE REVIEW

**Review by Strohgelaender — APPROVED** [review]
code lgtm

**Review by JohannesStoehr — DISMISSED** [review]
Approve code changes

**Review by MichaelOwenDyer — DISMISSED** [review]
Looks good to me 👍🏽

**Review by laadvo — DISMISSED** [review]
tested on ts1, the exercise group name is shown to students in breadcrumbs for the submission and the example solution

File: src/main/java/de/tum/in/www1/artemis/service/TitleCacheEvictionService.java [hunk 1]
```diff
@@ -100,6 +102,16 @@ else if (entity instanceof ExerciseHint hint) {
...
             evictIdFromCache("exerciseHintTitle", combinedId);
         }
+        else if (entity instanceof ExerciseGroup exerciseGroup) {
+            if (!Hibernate.isInitialized(exerciseGroup.getExercises())) {
+                log.warn("Unable to clear title of exercises from exercise group {}: Exercises not initialized", exerciseGroup.getId());
+                return;
+            }
+            var exercises = exerciseGroup.getExercises();
+            for (Exercise exercise : exercises) {
+                evictIdFromCache("exerciseTitle", exercise.getId());
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/TitleCacheEvictionService.java:112]: why is this necessary?

Comment (tobias-lippert) [reply src/main/java/de/tum/in/www1/artemis/service/TitleCacheEvictionService.java:112]: We should remove the title from the cache if the title changes to not get outdated cache values.

**Review by terlan98 — APPROVED** [review]
Tested on TS5. Works as expected. Students only see the exercise group name in the breadcrumb while tutors see the exercise name 👍

**Review by bassner — CHANGES_REQUESTED** [review]
I'm not happy with this solution.

These changes cause the original single database query (or, zero DB queries for most students as it's cached after the first request) to fetch an exercise title to now need 2-3 DB queries for a normal student just to return an entity title (load the exercise, load the user with groups, then load the title of the group). This entire thing was designed so that if the title has to actually be downloaded for each student (which will then maybe cause 2000 HTTP requests at once) it will not cause much load. That's why we aggressively cache on all ends.

Since you already have the correct title to display in #6873, I'm wondering why it is not possible to achieve this with a pure client side change. 
Alternatively, you could decide on the client wether you want to load the name of the exercise or the name of the exercise group for the breadcrumb; the data for that decision (user's groups, exercise group data, exercise data) is available in the client.

**Review by JohannesStoehr — APPROVED** [review]
Approve code. DB calls should not be more than before

**Review by terlan98 — COMMENTED** [review]
Retested on TS5. I've noticed that when viewing a result as a student, the breadcrumb initially shows `exercises` before showing the exercise group name. Can we remove this delay somehow?

https://github.com/ls1intum/Artemis/assets/22798314/990aad4e-be12-4805-b0fc-64e2cc755158

**Review by MaximilianAnzinger — APPROVED** [review]
Code LGTM

**Comment by tobias-lippert** [comment]
> Retested on TS5. I've noticed that when viewing a result as a student, the breadcrumb initially shows `exercises` before showing the exercise group name. Can we remove this delay somehow?
> 
>  bug_recording.mov

No, we cannot, if you are the first student who retrieves the exercise title for the breadcrumb. In all subsequent cases it will be much faster.

**Review by FelixTJDietrich — APPROVED** [review]
Tested on ts5 with a quiz, described changes work as expected.

**Review by pal03377 — APPROVED** [review]
Tested on TS5 in the exam testing session. Everything worked as described 🙈

**Review by RY997 — APPROVED** [review]
Tested on ts5 in test session, worked well

**Review by basak-akan — APPROVED** [review]
Tested on ts5 in test session worked as expected.
